src/librustc_typeck/check/op.rs

```diff
@@ -253,6 +254,50 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 // error types are considered "builtin"
                 if !lhs_ty.references_error() {
                     let source_map = self.tcx.sess.source_map();
+
+                    let suggest_constraining_param =
+                        |mut err: &mut DiagnosticBuilder<'_>,
+                         missing_trait: &str,
+                         p: ty::ParamTy,
+                         set_output: bool| {
+                            let hir = self.tcx.hir();
+                            let msg =
+                                &format!("`{}` might need a bound for `{}`", lhs_ty, missing_trait);
+                            if let Some(def_id) = hir
```

Comment:
Nit: I think a comment would be nice here. I think something like 

---

Try to find the def-id and details for the parameter `p`. We have only the index,
so we have to find the enclosing function's def-id, then look through its declared
generic parameters to get the declaration.

---

Come to think of it, though, we have the `HirId` of the *body* being checked, would it be easier to invoke [`body_owner_def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.body_owner_def_id) on `self.body_id`?

Reply:
I used `body_id`, but it is a `hir::HirId`. I only wrapped it with a ` hir::BodyId` but a slightly larger change would be to change `body_id` to be a `BodyId` instead, but I didn't want to bloat this PR for the reviewer's sake :)